### ISSUE TITLE
Allow run mutation tests against same project

### ISSUE THREAD

**Issue by jedielson** [stryker-mutator/stryker-net#1767]
**Is your feature request related to a problem? Please describe.**
No.

**Describe the solution you'd like**
Given a lib, where the code and tests lives together, As a developer, I'd like to use mutation tests in same lib where my tests live.

**Describe alternatives you've considered**
In .NET ecossystem, we usually have a project for our prod code and another project that contains the tests for this code.

Other languages, like golang, js, and rust, allow testing code to live together with prod code.
Althougth in .Net it's not recommended, it's easy doing this, only by changing a csproj like this:

``` xml
<ItemGroup Condition="'$(Configuration)' != 'Release'">
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.9.4" />
    <PackageReference Include="xunit" Version="2.4.1" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.3">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
    <PackageReference Include="coverlet.collector" Version="3.0.2">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
  </ItemGroup>
```

But if you try stryker against this project, it won't run, because, maybe there's no reference, or maybe, there's more than one reference.

While I recognize that this is controversial (in .NET ecossystem), having tests side-by-side with encourages devs to build smaller methods and keep methods as pure as possible. 

There's the `private` problem too. Oftentimes, a method is not really being tested, because it's logic is too encapsulated behind lots of private methods.
We could use internal methods for this, using `InternalsVisibleTo` in some classes, but when a test method is put closer, devs write more, and better test code.

As I've said, there's a workaroud for doing this in .NET, but it's not possible use Striker for this case.

**Additional context**

**Comment by dupdob** [comment]
Thanks for this proposal.

TLDR;
this cannot be implemented without establishing some standard naming for this practice which are beyond the objectives of Stryker.

Putting aside wether this practice is good or not, this is practically impossible to implement (in the general case). Mutation testing relies on being able to distinguish between 'production code', that must be mutated, and 'test code' that must not be mutated. Having separate assemblies is obviously great for that. Stryker could try and identify test methods (thanks to the test attribute), but it could not identify test helper methods (i.e. methods used by tests while not being production code or a test fixture). Furthermore, the usage of test attributes is sometimes optional (as test runners leverage having test only assemblies).

There is the notion of 'modules' in .NET, but I never though anything leveraging those. So, this feature would require the user to specify some criteria helping to sort between production code and test code, maybe via dedicated namespaces. Note that for performance reason, Stryker currently does not build a semantic model of the code that would be required to identify namespace properly.

No, going back to the practice, I would like to stress out this is probably not a good idea in .NET as this implies distributing xUnit with your project. If your project is an application, this is simply extra dead weight, but if your project is a library, those are extra dependencies that will burden your library users.

If you need to test private code, there is a solution for that in .NET using [`[InternalsVisibleTo]`](https://docs.microsoft.com/en-us/dotnet/standard/assembly/friend) (and using `internal` instead of `private`). 

One way to move forward with this is for you to provide some pointers regarding standard practices (naming or other) related to keeping code and tests within the same assembly.

> _On a personal note, I would like to point that this is a frowned upon practice for very good reasons, but this is not the place to discuss that._

**Comment by jedielson** [comment]
Thanks for your answer @dupdob! 😄 

> Putting aside wether this practice is good or not, this is practically impossible to implement (in the general case). Mutation testing relies on being able to distinguish between 'production code', that must be mutated, and 'test code' that must not be mutated. Having separate assemblies is obviously great for that. Stryker could try and identify test methods (thanks to the test attribute), but it could not identify test helper methods (i.e. methods used by tests while not being production code or a test fixture). Furthermore, the usage of test attributes is sometimes optional (as test runners leverage having test only assemblies).

I agree this is a big issue. In those languages I've mentioned, there are built in features that make things workout.
My suggestion would be using a class name suffix, like `Test`. In golang and python, this is the criteria. 😅 

> No, going back to the practice, I would like to stress out this is probably not a good idea in .NET as this implies distributing xUnit with your project. If your project is an application, this is simply extra dead weight, but if your project is a library, those are extra dependencies that will burden your library users.

It is relativelly easy to resolve, just by putting something like this in .csproj. The drawback is that it has to be done by some developer, during project configuration.

``` xml
<ItemGroup Condition="'$(Configuration)' == 'Release'">
    <Compile Remove="**\*.Tests.cs" />
  </ItemGroup>
```

The Xunit itself is no distributed with your lib, because you have a condition for include this libs only in a non-release build.

``` xml
<ItemGroup Condition="'$(Configuration)' != 'Release'">
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.9.4" />
    <PackageReference Include="xunit" Version="2.4.1" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.3">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
    <PackageReference Include="coverlet.collector" Version="3.0.2">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
  </ItemGroup>

```

I get your point that all the standardization part is really out of our concerns here.
The '.NET way' is putting test files into a separate lib. As you said, this is not the place to discuss it, but is really nice have test code side by side with prod code.
Do you know how it's done in js version of Stryker?

Once more, thanks for your help.

**Comment by rouke-broersma** [comment]
Building a custom test case resolver is IMO not within the scope of stryker.net and for that reason I doubt we will want to support this scenario. Even if this was contributed by a 3rd party I unfortunately don't think we will merge such a pull request because frankly I don't want to take up the burden of supporting such unorthodox functionality and if in the future it breaks we will probably have to fix it.

I realize we will be excluding some users with this decision but tbf I don't think the number of users with this wish are large enough to warrant supporting this. If you are a user that wants this functionality please let us know on this issue so we can see if I'm wrong about that.
